libavcodec/hevcdec.c

```diff
@@ -2631,6 +2630,7 @@ static int hls_slice_data_wpp(HEVCContext *s, const H2645NAL *nal)
     for (i = 1; i < s->threads_number; i++) {
         s->sList[i]->HEVClc->first_qp_group = 1;
         s->sList[i]->HEVClc->qp_y = s->sList[0]->HEVClc->qp_y;
+        s->sList[i]->HEVClc->gb = s->sList[0]->HEVClc->gb;
```

Comment:
Why do we have to cache `gb` here?

Reply:
Good question.

The idea is this - when we start to decode a wpp row with hls_decode_entry_wpp(), there are 2 parameters we pass in that are relevant here:
- Which thread we're using (0 -> n threads) - this is called self_id
- Which row we're processing - this is called job
- The thread and the job aren't necessarily related - thread 0 might start on row 3, and thread 5 might start on row 0, etc.

Furthermore the "gb" there stands for "get bits" - it holds the position in the bitstream.

The problem was:
- before this change, the 0th local context object was already set to the starting position of the first row in the slice.  So pulling from the bit stream would get you row 0 data.
- however, the 0th local context object is not guaranteed to be used for row 0 - it's guaranteed to be used for _thread_ 0.  So if thread 0 doesn't have row 0, this breaks.
- it so happens that the ffmpeg slice threading implementation does guarantee (probably by accident) that the first thread processes the first row.  After each thread grabs a row, further row to thread assignemnts are random, depending which thread finishes work first.
- but - I have plans to use our own thread pool to do intra-frame multithreading, which has no such guarantee.  So I want to set up every local context object at the start of the slice so that whichever one starts on row 0 it'll be pointed at the right location.